src/drivers/intel/cavs/hda-dma.c

```diff
@@ -403,8 +403,20 @@ static int hda_dma_link_copy_ch(struct dma *dma, struct hda_chan_data *chan,
 
 	/* make sure that previous transfer is complete (playback only) */
 	if (chan->direction == DMA_DIR_MEM_TO_DEV) {
-		while (hda_dma_get_free_size(dma, chan->index) < bytes)
+		uint64_t deadline = platform_timer_get(platform_timer) +
+			clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) *
+					  PLATFORM_LINK_DMA_TIMEOUT / 1000;
+
+		while (hda_dma_get_free_size(dma, chan->index) < bytes) {
+			if (deadline < platform_timer_get(platform_timer)) {
+				trace_hddma_error("hda-dmac: %d link timeout",
+						  dma->plat_data.id);
+				/* TODO: Timeout handling? */
+				return -ETIME;
+			}
```

Comment:
I would make this a static inline function e.g. hda_wait_for_transfer_complete() as it could also be used in other places in the future and makes the logic a little more readable.